### PR TITLE
Configure Dockerfile + CircleCI to correctly build node + protoc files needed for GUI

### PR DESCRIPTION
**Pull request by mark-idleman**
Most of the relevant action here is in `Dockerfile.model`, which is what we use to build our base GraphHopper docker images. The main changes are properly installing `npm` and `protoc` so the steps that set up Node components + generate Java/JS code via proto files will actually run correctly, and massively reorganizing the dockerfile into a many-stage build to make all the various build steps (each requiring different build tooling) will run without issues.

I also turned on docker layer caching in Circle, as it seems like this is the right thing to do given how many steps we have in our dockerfile now. There are probably small opportunities to condense some of these steps into shell scrips to reduce the number of docker build steps that need to be run (similar to what we did with `./build_with_protos` in the dockerfile previously), but I'll leave those changes to later PRs, and keep things verbose (and clear) for now

Proof that this works: I used the base tag created from these changes to build norcal routers, which are now running w/ GUI in GKE: https://replicahq.slack.com/archives/CK358RL0Z/p1619195617099800?thread_ts=1618595994.082700&cid=CK358RL0Z

### CODE REVIEW

File: Dockerfile.model [hunk 1]
```diff
@@ -1,28 +1,68 @@
...
+    && mkdir -p ./grpc/src/main/resources/assets/pt/src/grpc
+RUN protoc grpc/src/main/proto/router.proto --js_out=import_style=commonjs:grpc/src/main/resources/assets/pt/src/grpc --grpc-web_out=import_style=commonjs,mode=grpcwebtext:grpc/src/main/resources/assets/pt/src/grpc
+
+
+# Build node modules
+FROM node:14-alpine as node-build
+
+WORKDIR /graphhopper
+COPY . .
+RUN rm -r grpc
```
Comment (JacobHayes) [inline Dockerfile.model:32]: Think we can remove this too if we're more precise with the `COPY`

File: Dockerfile.model [hunk 2]
```diff
@@ -1,28 +1,68 @@
...
 ARG GRAPHHOPPER_TOKEN
 ARG GIT_USER_ID
 ARG GIT_PERSONAL_ACCESS
 
-RUN apt-get update && apt-get install -y wget
-
 WORKDIR /graphhopper
-
 COPY . .
+RUN rm -r grpc
```
Comment (JacobHayes) [inline Dockerfile.model:47]: ditto, though I'm guessing this actually does need more stuff

File: Dockerfile.model [hunk 3]
```diff
@@ -1,28 +1,68 @@
...
+COPY --from=node-build ./graphhopper/grpc ./grpc
+
+RUN mvn -s maven_settings.xml --projects grpc -am -DskipTests=true package \
+    && mvn -s maven_settings.xml --projects web -am -DskipTests=true package
```
Comment (JacobHayes) [inline Dockerfile.model:51]: nit: I usually prefer the `&&` at the end:
```suggestion
RUN mvn -s maven_settings.xml --projects grpc -am -DskipTests=true package && \
    mvn -s maven_settings.xml --projects web -am -DskipTests=true package
```

**Review by JacobHayes — COMMENTED** [review]
Nice!

**Review by JacobHayes — APPROVED** [review]
Oops, meant to "approve"

File: Dockerfile.model [hunk 4]
```diff
@@ -1,28 +1,68 @@
...
+    && cp ./idls/model/router.proto ./grpc/src/main/proto/router.proto \
+    && mkdir -p ./grpc/src/main/resources/assets/pt/src/grpc
+RUN protoc grpc/src/main/proto/router.proto --js_out=import_style=commonjs:grpc/src/main/resources/assets/pt/src/grpc --grpc-web_out=import_style=commonjs,mode=grpcwebtext:grpc/src/main/resources/assets/pt/src/grpc
+
+
+# Build node modules
+FROM node:14-alpine as node-build
+
+WORKDIR /graphhopper
+COPY . .
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply Dockerfile.model:31]: Not a fan of my lazy COPYing? ;) Our fork of the GH repo is very pared down already, so my thought process was that doing this step imprecisely wouldn't really hurt us; the whole repo, uncompressed, is ~400kb, so I'm happy to update these COPYs if we really think caching will improve, but I'm not sure going from 400 to 200/100kb will really make much difference, right?

re: the two `package.json`s, the one inside the `web` folder is basically defining a "local" package that points to a bunch of actual web assets in the `grpc` folder; the top-level `package.json` then points to this local package (see [this line](https://github.com/replicahq/graphhopper/pull/62/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R9)). I'm a bit of a Node noob, and this code was basically written by 2 Java developers haha, but @devthiago gave guidance here and everything seems to work as-is, so I'm scared of touching it too much 😭

[the remark this answers is not in the document]
Comment (JacobHayes) [reply Dockerfile.model:31]: It's not a matter of the size but invalidating the downstream build steps - any change to any file will cause the whole npm step to rerun (which also triggers all `FROM` downstream of that). Docker (understandably) is not quite like arti where it only checks if the _output_ didn't change (and thus can cache downstream), it only looks if the _input_ changes and will invalidate all else if so.

[the remark this answers is not in the document]
Comment (mark-idleman) [reply Dockerfile.model:31]: Ah good point, forgot how the caching worked exactly. I'll mess w/ slimming down the `COPY`s now

File: Dockerfile.model (hunk 2 above)
Comment (mark-idleman) [reply Dockerfile.model:47]: This needs basically everything except the `grpc` folder....I think I'm gonna keep this one for the sake of getting this in ASAP

**Pull request merged**